### DISCUSSION TITLE
Change author list

### DISCUSSION THREAD

**Discussion by xju2** [retorquere/zotero-better-bibtex#2822] [Q&A] [answered]
I want to use the postscript feature to change the author list in case the paper is from a collaboration such as the ATLAS Collaboration https://atlas.cern/. Such a paper will contain thousands of authors, e.g. https://inspirehep.net/literature/1685420. What would be the javascript snippets to do that? 

I tried the following but it did not work:

```javascript
if (Translator.BetterTeX) {
  if (tex.has.author && tex.author.includes("ATLAS Collaboration")) {
    tex.remove("author");
    tex.add({name: "author", value: "ATLAS Collaboration"});
  }

  if (tex.has.author && tex.author.includes("CMS Collaboration")) {
    tex.remove("author");
    tex.add({name: "author", value: "CMS Collaboration"});
  }
}
```

**Comment by retorquere** [comment]
I'll need a debug log.

**Reply by xju2** [reply]
[USJQT632-apse.tgz](https://github.com/retorquere/zotero-better-bibtex/files/14749472/USJQT632-apse.tgz)
Please let me know should you need more information.

**Reply by retorquere** [reply]
I need one generated by right-clicking the entry with the many authors and sending a log from the popup menu. The ID will have `-refs-` in it. I prefer the debug log just being sent vs saving and attaching it here.

**Reply by xju2** [reply]
Thanks for the instructions. The debug log ID: `Q3FP2VD8-refs-euc/6.7.172-6`.

**Reply by retorquere** [reply]
this will do it:

```
if (Translator.BetterTeX) {
  if (tex.has.author && tex.has.author.bibtex.includes("ATLAS Collaboration")) {
    // the "new String" ensures proper quoting, since you're doing it manually
    tex.add({name: "author", value: new String("ATLAS Collaboration")}); 
  }

  if (tex.has.author && tex.has.author.bibtex.includes("CMS Collaboration")) {
    tex.add({name: "author", value: new String("CMS Collaboration")});
  }
}
```

but I'd recommend this (handles proper quoting automatically):

```
if (Translator.BetterTeX) {
  function collaboration(name) {
    const authors = zotero.creators.filter(cr => cr.creatorType == 'author')
    // 5 is arbitrary, and if you always have it replaced, just remove that check entirely
    if (authors.length > 5 && authors.find(cr => cr.name == name)) {
      zotero.creators = zotero.creators.filter(cr => cr.creatorType != 'author' || cr.name == name)
      return true
    }
  }

  if (collaboration("ATLAS Collaboration") || collaboration("CMS Collaboration")) {
    tex.addCreators()
  }
}
```

**Reply by xju2** [reply, marked as answer]
Thanks for the prompt replies. I appreciate it. 

The two collaborations often have > 1000 people in the author list. I use a name to identify if the paper belongs to the collaboration as in some cases the "ATLAS/CMS Collaboration" does not exist in the author list. It should work until the person leaves the collaboration.

I am not familiar with javascript. The following works for the collection of papers I have.

```javascript
if (Translator.BetterTeX) {
  const authors = zotero.creators.filter(cr => cr.creatorType == 'author')

  if (authors.length > 1000 && tex.has.author && tex.has.author.bibtex.includes("Ju, X.")) {
    tex.add({name: "author", value: new String("ATLAS Collaboration")}); 
  }

  if (authors.length > 1000 && tex.has.author && tex.has.author.bibtex.includes("Sirunyan, A.")) {
    tex.add({name: "author", value: new String("CMS Collaboration")});
  }
}
```

**Reply by retorquere** [reply]
If that does what you want that's fine.

**Reply by xju2** [reply]
Thanks for the help!
